### ISSUE TITLE
Enable HttpClient (and other Micronaut internals) logging with correlation ID (OpenTracing + Logging support)

### ISSUE THREAD

**Issue by stepanv** [micronaut-projects/micronaut-core#4603]
Hi Micronaut team,

I'd like to discuss a possible better support of a propagation of "some" correlation ID to the log messages in Micronaut.

TL;DR 
Ideally, to better troubleshoot what is going inside Micronaut application under a load of multiple users in parallel, we would like to have in all log messages some correlation ID so that it's clear what log messages belong to each other. All log messages logged during a processing of a single incoming http request should contain the same correlation ID. This correlation ID should be propagated over network to upstream services as well (OpenTracing).

To achieve this, we 

- use Micronaut `ServerRequestContext` to store that correlation ID (among with the principal which is provided by Micronaut)
- we use Log4j2 `ContextDataInjector` to extract needed information from the `ServerRequestContext` (such as the correlation ID + principal) and enhance the log message 
(this is basically MDC with the difference that we use ThreadLocal of the ServerRequestContext instead of the MDC thread local - which is more reliable as it's Micronaut responsibility to correctly propagate ServerRequestContext)

the resulting log message may look like
```
2020-11-24 CET 22:23:26,569 [main] [correlationID=0720f66c4b864ae39fe00b71f88601e0] [principal=test@appd] INFO  com.appdynamics.query.service.DiagnosticContextComponentTest - HELLO WORLD
```
The `0720f66c4b864ae39fe00b71f88601e0` correlation ID enables me to search for the log messages that belong to a single user request even across multiple services in our complex system.

Anytime, we use logger in our "application" code using Micronaut, since the ServerRequestContext is correctly propagated, the resulting log messages are exactly as we need.

**However,**
when Micronaut logs, sometimes, it doesn't populate the ServerRequestContext and the logging subsystem has no idea what context/request we're currently in.
This is especially frustrating in case of the DefaultHttpClient debug and trace logs. When we need to trace what was the client request/response payload we get log messages without the correlation ID:
```
2020-11-24 CET 22:53:02,498 [multithreadEventLoopGroup-3-1] [correlationID=] [principal=] DEBUG io.micronaut.http.client.netty.DefaultHttpClient - Sending HTTP Request: GET /test
2020-11-24 CET 22:53:02,498 [multithreadEventLoopGroup-3-1] [correlationID=] [principal=] DEBUG io.micronaut.http.client.netty.DefaultHttpClient - HTTP Client Response Received for Request: GET http://localhost:51068/test
```
.

**To fix this,** 
the implementation of the DefaultHttpClient could always wrap the log statements into the `ServerRequestContext.with()` such as:
```
ServerRequestContext.with(request, () -> {
        if (log.isDebugEnabled()) {
            log.debug("Sending HTTP Request: {} {}", nettyRequest.method(), nettyRequest.uri());
            log.debug("Chosen Server: {}({})", requestURI.getHost(), requestURI.getPort());
        }
        if (log.isTraceEnabled()) {
            traceRequest(finalRequest, nettyRequest);
        }
});
```

This way a particular use-case would be supported. Ideally, Micronaut could "strictly" explicitly forward the request context anytime it's not inside the ServerRequestContext.with() already so that the Open Tracing + Logging use case is better supported. But I understand that would be quite tedious work.

What do you guys think about this? I can definitely submit a PR improving those few lines in `DefaultHttpClient`.

**Comment by graemerocher** [comment]
Seems reasonable. PRs welcome

**Comment by stepanv** [comment]
I created a PR https://github.com/micronaut-projects/micronaut-core/pull/4885 . I'd like to hear what you guys think about this approach.
It's definitely not ideal because the propagation of the ServerRequestContext is not "transparent" for the implementers of the `DefaultHttpClient`. Unfortunately, it's not possible to extend Netty handler factories in a way that it automagically adds the `ServerRequestContext` every time it's available. That would require a PR to Netty to be able to define own handler factory (and also improve the Netty code in a way that such factory is used always).

cc @graemerocher

**Comment by anuram28** [comment]
@stepanv could you please share how to are using ServerRequestContext & ContextDataInjector to log the correlation ID in all logs. Please a basic working example of how to use this in micronaut project would be very helpful. There is no documentation on this and any help on this greatly appreciated.

**Comment by stepanv** [comment]
Hi @anuram28 , I don't have a working example I could share right away. Feel free to create a minimal example by yourself and I shall be able to give you few hints if it wasn't working for you.
You need:
1. a java/kotlin project with slf4j + log4j2 + micronaut
3. configure logging to enable `TRACE` for `com.micronaut`
4. add a logging pattern that includes a reference to attribute `my_correlation_id`
5. fill this attribute `my_correlation_id` from the `ServerRequestScope` in the subtype of `ContextDataInjector` as described at https://github.com/micronaut-projects/micronaut-core/pull/4885

**Comment by crdhilep** [comment]
Sample code snippet 

> fun traceId(): String {
  var traceId: String? = UUID.randomUUID().toString()
    var currentRequest = ServerRequestContext.currentRequest<Any>()
      currentRequest.ifPresent {
      traceId = it.headers.get("traceId")
    } 
  return traceId.toString()
}
